lantern_embeddings_core/src/lib.rs

```diff
@@ -394,10 +396,115 @@ pub mod clip {
         Ok(())
     }
 
+    fn clear_model_cache(
+        model_map: &mut HashMap<&'static str, ModelInfo>,
+    ) -> Result<(), anyhow::Error> {
+        for (_, model_info) in model_map.iter_mut() {
+            model_info.encoder = None;
+        }
+
+        Ok(())
+    }
+
+    fn check_available_gpu_memory(mem_threshold: f64) -> Result<bool, anyhow::Error> {
+        let mut _nvml_instance = None;
+        let mut gpu_device = None;
+
+        if let Ok(nvml) = Nvml::init() {
+            _nvml_instance = Some(nvml);
+            let _nvml_insteance = _nvml_instance.as_mut().unwrap();
+            let nvml_device = _nvml_insteance.device_by_index(0);
+            if let Ok(device) = nvml_device {
+                gpu_device = Some(device);
+            }
+        }
+
+        if gpu_device.is_none() {
+            return Ok(true);
+        }
+
+        let gpu_device = gpu_device.as_ref().unwrap();
+
+        let mem_info = gpu_device.memory_info()?;
+        // Sometimes models consume much more memory based on input size.
+        // In my experiments the 200mb model was consuming 3GB memory.
+        // So comparing model size with free GPU memory will not be correct here.
+        // Instead we will check if the used memory in GPU is more than the threshold
+        // For GPU we will just clear the cache but not throw an error
+        // As the error from ORT is handled and will not cause an OOM like RAM
+        let percent_of_free_mem = (mem_info.used as f64 / mem_info.total as f64) * 100.0;
+
+        if percent_of_free_mem >= mem_threshold {
+            return Ok(false);
+        }
+        return Ok(true);
+    }
+
+    fn check_available_memory(
+        logger: &LoggerFn,
+        model_path: &PathBuf,
+        model_map: &mut HashMap<&'static str, ModelInfo>,
+        cache: bool,
+    ) -> Result<(), anyhow::Error> {
+        let mut sys = System::new_all();
+        sys.refresh_all();
+        let total_free_mem =
+            (sys.total_memory() - sys.used_memory()) + (sys.total_swap() - sys.used_swap());
+        let total_free_mem = total_free_mem as f64;
+        let model_file = std::fs::File::open(model_path)?;
+        let metadata = model_file.metadata()?;
+        let model_size = metadata.len() as f64;
+
+        let percent_of_free_mem = (model_size / total_free_mem) * 100.0;
+        let mem_threshold: f64 = 80.0;
```

Comment:
Move this to somewhere more visible, near other constants, in case we want to change it